### PR TITLE
[SPARK-36245][SQL] Deduplicate the right side of left semi/anti join

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Add a new rule to deduplicate the right side of left semi/anti join if it cannot be planed as broadcast hash join and there are many duplicate values. For example:
```scala
sql("CREATE TABLE t1 USING PARQUET AS SELECT id a, id AS b, id AS c FROM range(100000000)")
sql("CREATE TABLE t2 USING PARQUET AS SELECT id % 500000 AS a, id % 500000 AS b FROM range(50000000)")
sql("ANALYZE TABLE t1 COMPUTE STATISTICS FOR ALL COLUMNS")
sql("ANALYZE TABLE t2 COMPUTE STATISTICS FOR ALL COLUMNS")
sql("set spark.sql.cbo.enabled=true")
sql("SELECT * FROM t1 LEFT SEMI JOIN t2 ON t1.a = t2.a").explain("cost")
```
Before this pr:
```
== Optimized Logical Plan ==
Join LeftSemi, (a#11575L = a#11578L), Statistics(sizeInBytes=3.0 GiB, rowCount=1.00E+8)
:- Filter isnotnull(a#11575L), Statistics(sizeInBytes=3.0 GiB, rowCount=1.00E+8)
:  +- Relation default.t1[a#11575L,b#11576L,c#11577L] parquet, Statistics(sizeInBytes=3.0 GiB, rowCount=1.00E+8)
+- Project [a#11578L], Statistics(sizeInBytes=762.9 MiB, rowCount=5.00E+7)
   +- Filter isnotnull(a#11578L), Statistics(sizeInBytes=1144.4 MiB, rowCount=5.00E+7)
      +- Relation default.t2[a#11578L,b#11579L] parquet, Statistics(sizeInBytes=1144.4 MiB, rowCount=5.00E+7)
```
After this pr:
```
== Optimized Logical Plan ==
Join LeftSemi, (a#11575L = a#11578L), Statistics(sizeInBytes=3.0 GiB, rowCount=1.00E+8)
:- Filter isnotnull(a#11575L), Statistics(sizeInBytes=3.0 GiB, rowCount=1.00E+8)
:  +- Relation default.t1[a#11575L,b#11576L,c#11577L] parquet, Statistics(sizeInBytes=3.0 GiB, rowCount=1.00E+8)
+- Aggregate [a#11578L], [a#11578L], Statistics(sizeInBytes=7.2 MiB, rowCount=4.70E+5)
   +- Project [a#11578L], Statistics(sizeInBytes=762.9 MiB, rowCount=5.00E+7)
      +- Filter isnotnull(a#11578L), Statistics(sizeInBytes=1144.4 MiB, rowCount=5.00E+7)
         +- Relation default.t2[a#11578L,b#11579L] parquet, Statistics(sizeInBytes=1144.4 MiB, rowCount=5.00E+7)
```

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and TPC-DS benchmark test.

1. spark.sql.cbo.enabled=true and spark.sql.cbo.joinReorder.enabled=true

SQL | Before this PR(Seconds) | After this PR(Seconds)
-- | -- | --
q10 | 15  | 10
q14a | 174 | 78
q35 | 34 | 23
q69 | 14 | 8
q95 | 38 | 26

2. spark.sql.cbo.enabled=false

SQL | Before this PR(Seconds) | After this PR(Seconds)
-- | -- | --
q10 | 20  | 11
q35 | 38 | 29
q69 | 27 | 13

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * Deduplicate the right side of left semi/anti join if it cannot be planed as broadcast hash join
+ * and there are many duplicate values.
+ */
+object DeduplicateRightSideOfLeftSemiAntiJoin extends Rule[LogicalPlan] with JoinSelectionHelper {
+  def apply(plan: LogicalPlan): LogicalPlan =
+    plan.transformWithPruning(_.containsAnyPattern(LEFT_SEMI_OR_ANTI_JOIN)) {
+      case j @ Join(_, right, LeftSemi | LeftAnti, _, hint)
+        if conf.cboEnabled && !canBroadcastBySize(right, conf) && !hintToBroadcastRight(hint) &&
+          muchSmaller(Distinct(right), right) =>
+        j.copy(right = Aggregate(right.output, right.output, right))
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:35]: How does this deduplicate the right side?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoinSuite.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            }
+
+            comparePlans(Optimize.execute(query), correctAnswer)
+          }
+        }
+      }
+    }
+  }
+
+  test("Should not deduplicate the right side if can not reduce the data") {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoinSuite.scala:88]: Maybe we should add JIRA prefix

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.plans.{LeftAnti, LeftSemi}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Distinct, Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.LEFT_SEMI_OR_ANTI_JOIN
+
+/**
+ * Deduplicate the right side of left semi/anti join if it cannot be planed as broadcast hash join
+ * and there are many duplicate values.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:27]: I would explain more how this rule deduplicate. It looks just adding aggregation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:35]: Add a `Distinct` and replace distinct with aggregate. Similar to `ReplaceDistinctWithAggregate`:
https://github.com/apache/spark/blob/d89da31d57c6c52d28c35d175cd219cf6f1f13cd/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L1776-L1781

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ *   SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1=b1
+ *   ==>  SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN (SELECT b1 Tab2 GROUP BY b1) t2 ON a1=b1
+ * }}}
+ */
+object DeduplicateRightSideOfLeftSemiAntiJoin extends Rule[LogicalPlan] with JoinSelectionHelper {
+  def apply(plan: LogicalPlan): LogicalPlan =
+    plan.transformWithPruning(_.containsAnyPattern(LEFT_SEMI_OR_ANTI_JOIN)) {
+      case j @ Join(_, right, LeftSemi | LeftAnti, _, hint)
+        if conf.cboEnabled && !canBroadcastBySize(right, conf) && !hintToBroadcastRight(hint) &&
+          muchSmaller(Distinct(right), right) =>
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:38]: shall we define another `muchSmaller` here? It seems weird to call into join's `muchSmaller` method. btw wondering why by default 3x is a good reduction value? Intuitively it seems to me that if reduction ratio is 2x (reduce by half) is already very good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * and there are many duplicate values.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1=b1
+ *   ==>  SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN (SELECT b1 Tab2 GROUP BY b1) t2 ON a1=b1
+ * }}}
+ */
+object DeduplicateRightSideOfLeftSemiAntiJoin extends Rule[LogicalPlan] with JoinSelectionHelper {
+  def apply(plan: LogicalPlan): LogicalPlan =
+    plan.transformWithPruning(_.containsAnyPattern(LEFT_SEMI_OR_ANTI_JOIN)) {
+      case j @ Join(_, right, LeftSemi | LeftAnti, _, hint)
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:36]: I assume we only want to do the deduplication for equi-join right? Shall we check the `Join` to be `ExtractEquiJoinKeys`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Distinct, Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.LEFT_SEMI_OR_ANTI_JOIN
+
+/**
+ * Deduplicate the right side of left semi/anti join if it cannot be planed as broadcast hash join
+ * and there are many duplicate values.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN Tab2 ON a1=b1
+ *   ==>  SELECT a1, a2 FROM Tab1 LEFT SEMI JOIN (SELECT b1 Tab2 GROUP BY b1) t2 ON a1=b1
```
Comment (LantaoJin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:30]: You miss a `FROM` in `(SELECT b1 Tab2 GROUP BY b1)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala (hunk 5 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala:36]: Is there some issue if deduplication for nonEqui-join?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala (hunk 4 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala]: ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala (hunk 5 above)
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DeduplicateRightSideOfLeftSemiAntiJoin.scala]: I think you are right, after a second thought, it seems no issue to me. Thanks.

**Pull request closed without merging**